crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,397 @@
+            info!(
+                ?repository_latest_lsn,
+                new_snapshot_lsn = ?snapshot.min_applied_lsn,
+                "Newly created snapshot is not newer than the latest stored snapshot, will not update latest pointer"
+            );
+            return Err(anyhow!(
+                "Snapshot repository already contains snapshot at LSN {}",
+                repository_latest_lsn,
+            ));
+        }
+
+        let latest_json = PutPayload::from(serde_json::to_string_pretty(&latest)?);
+        let conditions = maybe_stored
+            .map(|(attributes, _)| PutOptions {
+                mode: PutMode::Overwrite,
+                attributes,
+                ..PutOptions::default()
+            })
+            .unwrap_or_else(|| PutOptions {
+                mode: PutMode::Create,
+                ..PutOptions::default()
+            });
+        let put_result = self
+            .object_store
+            .put_opts(&latest_path, latest_json, conditions)
+            .await?;
+        debug!(
+            etag = put_result.e_tag.unwrap_or_default(),
+            key = ?latest_path,
+            "Successfully updated latest snapshot pointer",
+        );
+
+        tokio::fs::remove_dir_all(local_snapshot_path.as_path()).await?;
+        trace!(
+            "Removed local snapshot files: {}",
+            local_snapshot_path.display()
+        );
+
+        Ok(())
+    }
+}
+
+/// S3 and other stores require a certain minimum size for the parts of a multipart upload. It is an
+/// API error to attempt a multipart put below this size, apart from the final segment.
+const MULTIPART_UPLOAD_CHUNK_SIZE_BYTES: usize = 5 * 1024 * 1024;
+
+async fn put_snapshot_object(
+    snapshot_path: &Path,
+    key: &object_store::path::Path,
+    object_store: &Arc<dyn ObjectStore>,
+) -> anyhow::Result<object_store::PutResult> {
+    let mut snapshot = tokio::fs::File::open(snapshot_path).await?;
+
+    if snapshot.metadata().await?.len() < MULTIPART_UPLOAD_CHUNK_SIZE_BYTES as u64 {
+        let payload = PutPayload::from(tokio::fs::read(snapshot_path).await?);
+        return object_store.put(key, payload).await.map_err(|e| e.into());
+    }
+
+    debug!("Performing multipart upload for {key}");
+    let mut upload = object_store.put_multipart(key).await?;
+
+    let mut buf = BytesMut::new();
+    let result: anyhow::Result<_> = async {
+        loop {
+            let mut len = 0;
+            buf.reserve(MULTIPART_UPLOAD_CHUNK_SIZE_BYTES);
+
+            // Ensure full buffer unless at EOF
+            while buf.len() < MULTIPART_UPLOAD_CHUNK_SIZE_BYTES {
+                len = snapshot.read_buf(&mut buf).await?;
+                if len == 0 {
+                    break;
+                }
+            }
+
+            if !buf.is_empty() {
+                upload
+                    .put_part(PutPayload::from_bytes(buf.split().freeze()))
```

Comment:
Getting the `split()` + `freeze()` / `reserve()` just right took me an embarrassingly long time! I have a reasonable grasp over `Bytes`/`BytesMut` now though, so that's a bonus. However please do review this carefully.

Something I wasn't 100% sure is whether it's possible for `AsyncReadExt::read_buf` to return less than a full buffer's worth. From what I can tell, it is possible - and S3 will throw an error if we try write a less than min chunk size part.